src/main/java/vazkii/botania/common/crafting/StateIngredientBlocks.java

```diff
@@ -37,6 +42,23 @@ public boolean test(BlockState state) {
 		return blocks.contains(state.getBlock());
 	}
 
+	@Override
+	public BlockState pick(Random random) {
+		if (blocks instanceof List) {
+			return ((List<Block>) blocks).get(random.nextInt(blocks.size())).getDefaultState();
+		}
+
+		// Slow path, iterating over the set
```

Comment:
could just store both the set and list forms so both `test` and `pick` are fast, at the cost of some ram. That would let you get rid of the extra ctor too.

Reply:
Looks like ImmutableSet has a nice method that converts it to a list (cached on the set), used that.